### PR TITLE
MooseBasedApp interface expansion

### LINKED ISSUE #540: MOOSE Input Parse Fails on Hanging Quote
--------
Issue Description
--------
##### What did you expect to see happen?
I was using Raven to do a parametric study for statistics on a MOOSE app. The MOOSE input file had a block that looked like this:
```
[Mesh]
  type = GeneratedBIDMesh # [cm]
  dim = 3
  xmin = 0.0
  xmax = 9.652
  ymin = 0.0
  ymax = 9.652
  zmin = 0.0
  zmax = 252.9320455
  nx = 2
  ny = 2
  nz = 23
  elem_type = HEX8
  subdomain = '
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  ' # 1 is outer, 2 is inside
  uniform_refine = 1
  1 Volume: 11269.4658501452 cm3
  2 Volume: 12293.962745613 cm3
[]
```
When Raven parsed the input file and did the code run, it messed up and put the " ' # 1 is outer, 2 is inside" line somewhere in the middle of the MOOSE block. The new block did not look like the original block and app had an error due to trying to convert parameter names to integers because it could not find the closing '.  

My **work around** was to change the code block to:
```
[Mesh]
  type = GeneratedBIDMesh # [cm]
  dim = 3
  xmin = 0.0
  xmax = 9.652
  ymin = 0.0
  ymax = 9.652
  zmin = 0.0
  zmax = 252.9320455
  nx = 2
  ny = 2
  nz = 23
  elem_type = HEX8
  subdomain = '
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  2 2 2 2
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1
  1 1 1 1'
  # 1 is outer, 2 is inside
  uniform_refine = 1
  1 Volume: 11269.4658501452 cm3
  2 Volume: 12293.962745613 cm3
[]
```
where the hanging ' is now on the line with the 1 1 1 1' instead of on the line with the comment. This worked just fine and the app ran and Raven did the job. The side note is that the MOOSE app is able to run both blocks just fine. Only when the original block went through Raven did the input block get messed up.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #794: MOOSE Input Incorrect Parse of Multi-line String
--------
Issue Description
--------
##### What did you expect to see happen?
In my MOOSE input file I have a multi-line string:
```
type = ParsedFunction
function = 'rT := temperature / 1000.0;
            cp_f := 52.1743 + 87.951 * rT - 82.2411 * pow(rT, 2) + 31.542 * pow(rT, 3) - 2.6334 * pow(rT, 4) - 0.7139 * pow(rT, -2);
            cp_rho_fuel := 1.0e-6 * rho_fuel / M_fuel * cp_f;
            cp_rho_graphite := 1e-9 * rho_graphite / (11.07 * pow(temperature, -1.644) + 0.0003688 * pow(temperature, 0.02191));
            rho_UO2 * cp_rho_fuel + (1 - rho_UO2) * cp_rho_graphite'
```
It treats each line as if it was not part of the string. This could be a problem if someone was changing a matrix of values using Raven.

##### What did you see instead?
The new input spawned by Raven had: (the first line had the problem in #792 )
```
function = 'rT :
type = ParsedFunction
cp_f : = 52.1743 + 87.951 * rT - 82.2411 * pow(rT, 2) + 31.542 * pow(rT, 3) - 2.6334 * pow(rT, 4) - 0.7139 * pow(rT, -2);
cp_rho_fuel : = 1.0e-6 * rho_fuel / M_fuel * cp_f;
cp_rho_graphite : = 1e-9 * rho_graphite / (11.07 * pow(temperature, -1.644) + 0.0003688 * pow(temperature, 0.02191));
rho_UO2 * cp_rho_fuel + (1 - rho_UO2) * cp_rho_graphite'
```
It doesn't preserve the order (which is important for a multi-line string) and some other lines in the block can end up between the lines that were part of the string.

##### Do you have a suggested fix for the development team?
Not really, I suggest to check for an open `'` and then end at the next `'`. I did find a work around: I could put the multi-line string on one line and it works (with the fix in #793 ). So perhaps taking a multi-line string and converting it to a single line string, first could work.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #540, Closes #794

##### What are the significant changes in functionality due to this change request?
Extends the robustness of the MooseBasedApp interface to handle a wider variety of inputs, including:
 - Multiline vectors
 - Scalar entries on different lines than their keys
 - Ordered inputs
 - Hanging vector opening/closing
 - Multiple "root" nodes in file

These are all tested via an interfaceOnly check with several challenging configurations that can be extended in the future for more testing and capability.

Also added is a resurrection of our getpot parser to input tree structure, used for reading and modifying the moose hit/Getpot input.

Note for anyone familiar with the old implementation, the concept of "attributes" has been done away in favor of treating all levels as nodes. This doesn't change anything by function, just mindset.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/CodeInterfaces/MooseBasedApp/MOOSEparser.py [hunk 1]
```diff
@@ -23,258 +23,103 @@
...
-    else:
-      if not found:
-        subElement = ET.SubElement(returnElement,name[0])
-        #if len(name) > 1, then if not found (and since we already checked for erasing) then add it and recurse.
-      else:
-        # if len(name) > 1 and found, then recurse on child
-        subElement = returnElement.find(true_name)
-      self.__modifyOrAdd(subElement,name[1:],modiDictionary)
+    # if asking to check on a block ...
+    # TODO is this untested code?
```
Comment (alfoa) [inline framework/CodeInterfaces/MooseBasedApp/MOOSEparser.py:107]: it looks like it is not tested...it was with Grizzly but we probably lost that test

**Comment by alfoa** [comment]
When the failed tests are fixed, we can merge it.
Not major comments on the PR itself.

**Comment by PaulTalbot-INL** [comment]
I broke the classicDETraven and hybridDETraven using the MOOSE-based file, but I'm not sure how I broke it.

Do you know what the trigger for the DET to end is, by chance @alfoa? It seems to go on sampling forever.

**Comment by joshua-cogliati-inl** [comment]
> I broke the classicDETraven and hybridDETraven using the MOOSE-based file, but I'm not sure how I broke it.
> 
> Do you know what the trigger for the DET to end is, by chance @alfoa? It seems to go on sampling forever.

I changed some parameters to get them to end sooner.  Now they seem to be passing (but there are other tests still failing)

**Comment by PaulTalbot-INL** [comment]
@alfoa tests passed

**Comment by alfoa** [comment]
Checklist is passed.

PR can be approved.

@PaulTalbot-INL can you email the user list informing about the fix?

Thanks a lot.

**Pull request merged**